org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/codemanipulation/OverrideMethodsOperation.java

```diff
@@ -0,0 +1,209 @@
+import org.eclipse.jdt.core.dom.AbstractTypeDeclaration;
+import org.eclipse.jdt.core.dom.AnonymousClassDeclaration;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.IMethodBinding;
+import org.eclipse.jdt.core.dom.IPackageBinding;
+import org.eclipse.jdt.core.dom.ITypeBinding;
+import org.eclipse.jdt.core.dom.MethodDeclaration;
+import org.eclipse.jdt.core.dom.Modifier;
+import org.eclipse.jdt.core.dom.rewrite.ASTRewrite;
+import org.eclipse.jdt.core.dom.rewrite.ImportRewrite;
+import org.eclipse.jdt.core.dom.rewrite.ImportRewrite.ImportRewriteContext;
+import org.eclipse.jdt.core.dom.rewrite.ListRewrite;
+import org.eclipse.jdt.internal.core.manipulation.StubUtility;
+import org.eclipse.jdt.internal.corext.codemanipulation.CodeGenerationSettings;
+import org.eclipse.jdt.internal.corext.codemanipulation.ContextSensitiveImportRewriteContext;
+import org.eclipse.jdt.internal.corext.codemanipulation.StubUtility2Core;
+import org.eclipse.jdt.internal.corext.dom.ASTNodes;
+import org.eclipse.jdt.internal.corext.dom.Bindings;
+import org.eclipse.jdt.internal.corext.dom.IASTSharedValues;
+import org.eclipse.jdt.internal.corext.refactoring.util.RefactoringASTParser;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+import org.eclipse.text.edits.MultiTextEdit;
+import org.eclipse.text.edits.TextEdit;
+
+public class OverrideMethodsOperation {
+
+	public static List<OverridableMethod> getOverridableMethods(IType type) {
+		if (type == null || type.getCompilationUnit() == null) {
+			return Collections.emptyList();
+		}
+
+		List<OverridableMethod> overridables = new ArrayList<>();
+		RefactoringASTParser astParser = new RefactoringASTParser(IASTSharedValues.SHARED_AST_LEVEL);
+		CompilationUnit astRoot = astParser.parse(type.getCompilationUnit(), true);
+		try {
+			ITypeBinding typeBinding = ASTNodes.getTypeBinding(astRoot, type);
+			if (typeBinding == null) {
+				return overridables;
+			}
+
+			IMethodBinding cloneMethod = null;
+			ITypeBinding cloneable = astRoot.getAST().resolveWellKnownType("java.lang.Cloneable");
+			if (Bindings.isSuperType(cloneable, typeBinding)) {
+				cloneMethod = resolveWellKnownCloneMethod(astRoot.getAST());
+			}
+
+			IPackageBinding pack = typeBinding.getPackage();
+			IMethodBinding[] methods = StubUtility2Core.getOverridableMethods(astRoot.getAST(), typeBinding, false);
+			for (IMethodBinding method : methods) {
+				if (Bindings.isVisibleInHierarchy(method, pack)) {
+					boolean toImplement = Objects.equals(method, cloneMethod);
+					overridables.add(convertToSerializableMethod(method, toImplement));
+				}
+			}
+		} catch (JavaModelException e) {
+			JavaLanguageServerPlugin.logException("List overridable methods", e);
+		}
+
+		return overridables;
+	}
+
+	public static TextEdit addOverridableMethods(IType type, OverridableMethod[] overridableMethods) {
```

Comment:
what happens if overridableMethods are actually already overridden (eg. call addOverridableMethods twice)?

Reply:
Generate duplicated code. Here i just take it as an utility method, and need the client to avoid calling the same code multiple times.